### ISSUE TITLE
Clients disconnected after upgrade to 0.7.0

### ISSUE THREAD

**Issue by thorstenkramm** [cloudradar-monitoring/rport#556]
Some users reported their clients got disconnected after the server has been updated to 0.7.0.

While the final error analysis is still pending, we could isolate the problem partially.
The recently introduced [SQLite WAL](https://sqlite.org/wal.html) mode for all databases, seems to delay or deny the availability of fresh data to all internal subsystems.

A client connects successfully, but the connection state is not updated. The API and a SQLite command line client are getting outdated data from the database.
We are still investigating and testing.

Meanwhile, do the following, and 🙏 please share your results:

1. Stop rportd and remove the clients.db.
  `mv /var/lib/rport/clients.db /var/lib/rport/clients.db.bak`
  Start rportd again. What happens? Are your clients flagged connected?
2. If the above does not help, stop rportd and disable Sqlite WAL in `rportd.conf`. 
  Remove the old database `mv /var/lib/rport/clients.db /var/lib/rport/clients.db.bak`
  Start rportd again. What happens? Are your clients flagged connected now?

*The clients.db is rebuilt automatically and server start.*

**Comment by amastelek** [comment]
I have tried to add a new client on 0.70 and its says established on the client side but its not in the web ui.   Must I follow the same process above?

**Comment by thorstenkramm** [comment]
Yes. Please reset the clientsdb as stated above.

**Comment by vagner-dias** [comment]
Hi. I tried both procedures above. On step 1, only some clients (around 30%) showed up in the interface. After the step 2, all clients disapeared.

**Comment by amastelek** [comment]
I did the procedure to reset the clientsdb.
All tunnels were wiped.
Clients not previously present would only show up when restarted using service rport restart.

**Comment by amastelek** [comment]
When I restart a client it goes red in the web ui and stays red until I restart the client with service rport restart

**Comment by vagner-dias** [comment]
Is there a way to roll back this update until this issue gets fixed?

**Comment by amastelek** [comment]
> Is there a way to roll back this update until this issue gets fixed?

Download the 0.6 version from release and replace  /usr/local/bin/rport

**Comment by amastelek** [comment]
BTW: Using debian buster and Raspbian.

**Comment by tscheiber5** [comment]
> 

I am experiencing the same issue with clients not showing as connected.
I've stopped rportd on the server renamed the clients.db and restarted rportd as suggested.
Unfortunately, only the clients that were connected before showed up.
I still need to connect to the other windows clients and restart the rport service manually to see them in the console.
Restarting a client service also worked before I restarted rportd on the server, so the database purge didn't solve the issue.

**Comment by vagner-dias** [comment]
Hi. I see one very consistent behaviour on this new version.

It happens on all the windows devices connected to rport.

Every time i restart a windows device, no mater the windows version (10, 7, 2012, 2019, 2022), it shows as offline (red) in rport console, so I have to log into the machine and restart the service manually. After that the PC shows as green in the console.

**Comment by thorstenkramm** [comment]
@vagner-dias Thank you for this hint. I can reproduce the behaviour you described on Windows. The log states the client to be connected. But on the server, it's shown disconnected.  We are investigating.

**Comment by thorstenkramm** [comment]
@vagner-dias Would you mind testing the attached version of rportd? 
Just replace `/usr/local/bin/rportd` and restart the service.
Please run the server with log level debug.

This version contains an extended connection logging. 
After a reboot the first connection takes longer than expected. The default timeout of 10 seconds is often exceeded.
The attached version has a timeout of 3 minutes. This solved the issue of my windows client. It now gets connected  reliably after a reboot. I assume that after the reboot the client collects inventory data and therefore the first connection takes longer. 
[rportd_0.7.1_linux_x86_64.gz](https://github.com/cloudradar-monitoring/rport/files/8855208/rportd_0.7.1_linux_x86_64.gz)

**Comment by amastelek** [comment]
> @amastelek Please look at the rport log. `/var/log/rport/rport.log` or `C:\Program Files\rport\rport.log`. `lsof` can just show a network connection. The client might be connected on network level but rejected due to bad credentials, fingerprint mismatch etc. The log will give you details.

I have a client on a raspberry and when I restart there is nothing logged in /var/log/rport/rport.log  No error, nothing.
I see this in service rport status:
● rport.service - Create reverse tunnels with ease.
   Loaded: loaded (/etc/systemd/system/rport.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2022-06-07 19:13:02 SAST; 7min ago
 Main PID: 1276 (rport)
    Tasks: 10 (limit: 4915)
   CGroup: /system.slice/rport.service
           └─1276 /usr/local/bin/rport -c /etc/rport/rport.conf

Jun 07 19:20:11 kl-speedtest sudo[2014]: pam_unix(sudo:session): session closed for user root
Jun 07 19:20:11 kl-speedtest sudo[2025]:    rport : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/usr/bin/docker inspect --format {{ .Name }} c3ab10ba0d3f3cb7769637bc49632
Jun 07 19:20:11 kl-speedtest sudo[2025]: pam_unix(sudo:session): session opened for user root by (uid=0)
Jun 07 19:20:11 kl-speedtest sudo[2025]: pam_unix(sudo:session): session closed for user root
Jun 07 19:20:11 kl-speedtest sudo[2036]:    rport : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/usr/bin/docker inspect --format {{ .Name }} c3ab10ba0d3f3cb7769637bc49632
Jun 07 19:20:11 kl-speedtest sudo[2036]: pam_unix(sudo:session): session opened for user root by (uid=0)
Jun 07 19:20:11 kl-speedtest sudo[2036]: pam_unix(sudo:session): session closed for user root
Jun 07 19:20:11 kl-speedtest sudo[2048]:    rport : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/usr/bin/docker inspect --format {{ .Name }} c3ab10ba0d3f3cb7769637bc49632
Jun 07 19:20:11 kl-speedtest sudo[2048]: pam_unix(sudo:session): session opened for user root by (uid=0)
Jun 07 19:20:11 kl-speedtest sudo[2048]: pam_unix(sudo:session): session closed for user root

Weird.  As I have a docker instance of openspeedtest running?

**Comment by amastelek** [comment]
On another debian buster client I have this in /var/log/rport/rport.log
2022/06/07 18:22:42 client: Update status refresh failed: sudo: a password is required

**Comment by thorstenkramm** [comment]
This must be a custom docker image because we don't release any. 
Having the log level set to `debug` the rportd.log logs all connection attempts and their status.

`tail -f /var/log/rport/rportd.log|grep "client-listener"`

```
2022/06/07 17:18:27 client-listener: client#259: Handshaking...
2022/06/07 17:18:27 client-listener: client#259: Verifying configuration...
2022/06/07 17:18:27 client-listener: client#259: Client version (0.5.20) differs from server version (0.7.1)
2022/06/07 17:18:27 client-listener: client#259: Failed: client id "37EA4D56-44A9-3381-CEC4-22D0E0A9D9B4" is already in use
2022/06/07 17:23:27 client-listener: client#260: Handshaking...
2022/06/07 17:23:28 client-listener: client#260: Verifying configuration...
2022/06/07 17:23:28 client-listener: client#260: Client version (0.5.20) differs from server version (0.7.1)
```

**Comment by vagner-dias** [comment]
@thorstenkramm , no much luck running this rportd version.

I tried running this version you provided on a ubuntu 20.04

rportd.service: Succeeded.
2022/06/07 17:25:56 listen tcp 0.0.0.0:80: bind: permission denied
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0xa38e09]
goroutine 55 [running]:
github.com/cloudradar-monitoring/rport/share.(*HTTPServer).Close(...)
        /var/www/rport/share/http_server.go:74
github.com/cloudradar-monitoring/rport/server.(*ClientListener).Close(0x0?)
        /var/www/rport/server/client_listener.go:167 +0x49
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /var/www/rport/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x67
created by golang.org/x/sync/errgroup.(*Group).Go
        /var/www/rport/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x8d

**Comment by thorstenkramm** [comment]
try `setcap CAP_NET_BIND_SERVICE=+eip /usr/local/bin/rportd` before starting the new binary.

**Comment by vagner-dias** [comment]
@thorstenkramm one interesting thing. If I change the rport windows service startup type to "automatic (delayed start)", everything works perfectly. 

Maybe this is related to the client and not to the server ?

**Comment by tscheiber5** [comment]
> @tscheiber5 Thank you for reporting what happened on your system. Just to clarify: A restart of the client re-connected them and they appeared in the client list, right?

@thorstenkramm yes, correct.

**Comment by ben-alphabeans** [comment]
> Hi. I see a very consistent behaviour on this new version.
> 
> It happens on all the windows devices connected to rport.
> 
> Every time i restart a windows device, no mater the windows version (10, 7, 2012, 2019, 2022), it shows as offline (red) in rport console, so I have to log into the machine and restart the service manually. After that the PC shows as green in the console.

I have a slightly different situation. In my network, the router reboots everyday. All windows clients disconnects after reboot.
Also, even if I set "cleanup_lost_clients = false", those client disappears.

The last line of the client log is about "Sending connection request..."

**Comment by tscheiber5** [comment]
I have some clients running with RPort 0.6.0 on Windows 10 Hosts.
The 0.6.0 has the same issue at the exact same time as the 0.7.0 clients.
![image](https://user-images.githubusercontent.com/26470855/172549083-74b6c796-cdb3-4f95-b1a4-06de500bbb09.png)

**Comment by thorstenkramm** [comment]
@tscheiber5 This supports my assumption the issue is caused by the server. Could you try [this update rportd_0.7.1_linux_x86_64.gz](https://github.com/cloudradar-monitoring/rport/files/8855208/rportd_0.7.1_linux_x86_64.gz) to check your clients are connecting more reliably?

**Comment by thorstenkramm** [comment]
> @thorstenkramm one interesting thing. If I change the rport windows service startup type to "automatic (delayed start)", everything works perfectly.
> 
> Maybe this is related to the client and not to the server ?

It supports my assumption that after a reboot the client needs more time to connect. Likely because the client is gathering the update status of the system. Internally, the update manager APIs of Windows are busy refreshing the data. That could slow down the rport client too. 
Increasing the connection timeout on the server seems to solve the problem. But as a future improvement, the client should first connect and send data about the update status later.

**Comment by tscheiber5** [comment]
> @tscheiber5 This supports my assumption the issue is caused by the server. Could you try [this update rportd_0.7.1_linux_x86_64.gz](https://github.com/cloudradar-monitoring/rport/files/8855208/rportd_0.7.1_linux_x86_64.gz) to check your clients are connecting more reliably?

@thorstenkramm I have updated the version of the server to 0.7.1 and will monitor the connectivity of the clients.

**Comment by ben-alphabeans** [comment]
I have tried rportd 0.7.1. Windows clients (0.7.0) cannot reconnect when the network is resumed
According to rport.log, the rport client disconnect when the network lost. Client then started retry attempts. In my PC, it take about 8 attempts to finally established a handshake. The log stop on "Sending connection request: {...}".
I think the server is able to respond with positive connection result that's why client stops retrying. However, the client is stated as disconnected on the server side.

**Comment by tscheiber5** [comment]
> @tscheiber5 This supports my assumption the issue is caused by the server. Could you try [this update rportd_0.7.1_linux_x86_64.gz](https://github.com/cloudradar-monitoring/rport/files/8855208/rportd_0.7.1_linux_x86_64.gz) to check your clients are connecting more reliably?

After a few days without any issue, I have some more disconnects of servers.
It only happened on one site where we had a short (about 10 minutes) internet outage.
Out of 9 PCs/servers 8 are running with 0.7.0 and one is running with 0.6.0.
All the PCs/servers are still on but showing as disconnected.
The only server that is still connected is running client version 0.6.0.

This is from the client log 0.7.0:
_2022/06/14 02:59:48 client: Monitoring ended by context.Done
2022/06/14 02:59:48 client: Connection error: read tcp 192.168.0.12:65223->xx.xx.xx.xx:80: wsarecv: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.
2022/06/14 02:59:48 client: Retrying in 100ms..._


I also have one server with client version 0.6.0 which is still connected.
here is the log file from version 0.6.0:
_2022/06/14 03:00:08 client: Connection error: read tcp 192.168.0.10:49226->xx.xx.xx.xx:80: wsarecv: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.
2022/06/14 03:00:08 client: Retrying in 100ms...
2022/06/14 03:00:08 client: Monitoring ended by context.Done
2022/06/14 03:00:09 client: Could not get OS Virtualization Info: exit status 0xfffd0000
2022/06/14 03:00:09 client: Connection error: client id "E39E1C67-8420-44A3-8853-9C12375758AA" is already in use (Attempt: 1)
2022/06/14 03:00:09 client: Retrying in 200ms...
2022/06/14 03:00:10 client: Could not get OS Virtualization Info: exit status 0xfffd0000
2022/06/14 03:00:10 client: Connection error: client id "E39E1C67-8420-44A3-8853-9C12375758AA" is already in use (Attempt: 2)
2022/06/14 03:00:10 client: Retrying in 400ms...
2022/06/14 03:00:11 client: Could not get OS Virtualization Info: exit status 0xfffd0000
2022/06/14 03:00:11 client: Connection error: client id "E39E1C67-8420-44A3-8853-9C12375758AA" is already in use (Attempt: 3)
2022/06/14 03:00:11 client: Retrying in 800ms...
2022/06/14 03:00:13 client: Could not get OS Virtualization Info: exit status 0xfffd0000
2022/06/14 03:00:13 client: Connection error: client id "E39E1C67-8420-44A3-8853-9C12375758AA" is already in use (Attempt: 4)
2022/06/14 03:00:13 client: Retrying in 1.6s...
2022/06/14 03:00:15 client: Could not get OS Virtualization Info: exit status 0xfffd0000
2022/06/14 03:00:15 client: Connection error: client id "E39E1C67-8420-44A3-8853-9C12375758AA" is already in use (Attempt: 5)
2022/06/14 03:00:15 client: Retrying in 3.2s...
2022/06/14 03:00:19 client: Could not get OS Virtualization Info: exit status 0xfffd0000
2022/06/14 03:00:19 client: Connection error: client id "E39E1C67-8420-44A3-8853-9C12375758AA" is already in use (Attempt: 6)
2022/06/14 03:00:19 client: Retrying in 6.4s...
2022/06/14 03:00:26 client: Could not get OS Virtualization Info: exit status 0xfffd0000
2022/06/14 03:00:26 client: Connection error: client id "E39E1C67-8420-44A3-8853-9C12375758AA" is already in use (Attempt: 7)
2022/06/14 03:00:26 client: Retrying in 12.8s...
2022/06/14 03:00:40 client: Could not get OS Virtualization Info: exit status 0xfffd0000
2022/06/14 03:00:40 client: Connection error: client id "E39E1C67-8420-44A3-8853-9C12375758AA" is already in use (Attempt: 8)
2022/06/14 03:00:40 client: Retrying in 25.6s...
2022/06/14 03:01:06 client: Could not get OS Virtualization Info: exit status 0xfffd0000
2022/06/14 03:01:06 client: Connection error: client id "E39E1C67-8420-44A3-8853-9C12375758AA" is already in use (Attempt: 9)
2022/06/14 03:01:06 client: Retrying in 51.2s...
2022/06/14 03:01:58 client: Could not get OS Virtualization Info: exit status 0xfffd0000
_

**Comment by amastelek** [comment]
I have a site that I'm unable to connect.

Here is the log:
2022/06/14 12:38:26 client: Update status refresh failed: sudo: a password is required
2022/06/14 12:41:02 client: data directory path: "/var/lib/rport"
2022/06/14 12:41:02 client: Connecting to ws://claymore.amastelek.com:8080
2022/06/14 12:41:02 client: Using AptPackageManager for updates
2022/06/14 12:41:02 client: Fingerprint 35:35:4d:15:b1:23:73:b0:58:55:3f:41:f9:ff:0e:96
2022/06/14 12:41:05 client: Update status refresh failed: context canceled

**Comment by thorstenkramm** [comment]
@tscheiber5 Can you try the following:
* Stop the rport server
* rename `clients.db` to something like `clients.db.bak`
* Start the server so the database is rebuilt from scratch.

Do you clients connect now?

**Comment by thorstenkramm** [comment]
I’m quite confident we untied the knot yesterday. 😎 
Thanks to @domenipavec  for the decisive hint.
We were able to reproduce the error on our test infrastructure. 

On client connection, the server applied the new “tunnel_allowed” policies to all clients, regardless their version. This failed on all clients < 0.6.4 and the clients could not connect.

This is now handled properly on server side. All clients – old and now ones – connect flawlessly and creating tunnels is not blocking.
You are kindly invited to test the attached version 0.7.3-rc1. 
**🙏 Feedback is highly appreciated.** 

[rportd_0.7.3-rc1_Linux_x86_64.tar.gz.md5.txt](https://github.com/cloudradar-monitoring/rport/files/8907551/rportd_0.7.3-rc1_Linux_x86_64.tar.gz.md5.txt)
[rportd_0.7.3-rc1_Linux_x86_64.tar.gz](https://github.com/cloudradar-monitoring/rport/files/8907553/rportd_0.7.3-rc1_Linux_x86_64.tar.gz)

**Comment by amastelek** [comment]
I'm getting clients to connect using the new binary.

**Comment by tscheiber5** [comment]
> @tscheiber5 Can you try the following:
> 
> * Stop the rport server
> * rename `clients.db` to something like `clients.db.bak`
> * Start the server so the database is rebuilt from scratch.
> 
> Do you clients connect now?

Unfortunately, only those clients connect that were connected before.
I've given them more then 24 hours but the servers (which I know are running) did not try to connect.
Only after a manual restart of the client service, they connect again.
So if the server (rportd 0.7.1) has an issue, it somehow affects the client service.

I've just upgraded to 0.7.3-rc-1, but the clients won't connect unless I restart the service.
On some of my clients including the server, I've changed the logging to debug in order to provide better information for you.

I'll monitor and keep you posted.

**Comment by ben-alphabeans** [comment]
My client version 0.6.0 for windows seems working fine with rportd 0.7.1.
I have tested it on two PC in my "daily disconnecting" network group.

**Comment by amastelek** [comment]
When there is an outage then client does not automatically reconnect.  Client 0.7 and server 0.7.3

**Comment by ffybots** [comment]
I'm having the same issue. On the server and client I have rport 0.7.0 and the allowed tunnel is not set in the server configuration.
I get this error on the client : 
> Monitoring ended by context.Done
> Update status refresh failed: Exception occurred. (The service cannot be started, either because it is disabled or because it has no enabled devices associated with it.
> )
> Could not get cpu information: context deadline exceeded

I reinstalled the previous version of rport on the client (didn't change it on the server) and it works fine now.

**Comment by thorstenkramm** [comment]
@ffybots 
We have released 0.8.0 stable meanwhile. This version should resolve the issue reliably.

**Comment by thorstenkramm** [comment]
@prjus I'm happy to help you.
Please re-open the issue [here](https://github.com/cloudradar-monitoring/rport-pairing/issues).
The update script has its own repo and issue tracker.
Please paste the full output of `Get-NetAdapter` executed on a PowerShell with admin-rights.
